source/common/ssl/tls_certificate_config_impl.cc

```diff
@@ -64,19 +64,42 @@ TlsCertificateConfigImpl::TlsCertificateConfigImpl(
           fmt::format("Certificate configuration can't have both pkcs12 and private_key_provider"));
     }
   } else {
-    if (config.has_private_key_provider()) {
+    if (config.has_private_key_provider_list()) {
+      for (const auto& provider : config.private_key_provider_list().private_key_provider()) {
+        private_key_method_ =
+            factory_context.sslContextManager()
+                .privateKeyMethodManager()
+                .createPrivateKeyMethodProvider(provider, factory_context,
+                                                const_cast<std::string&>(private_key_));
+        if (private_key_method_ && private_key_method_->checkInitialized()) {
+          break;
+        } else {
+          private_key_method_ = nullptr;
+        }
+      }
+    } else if (config.has_private_key_provider()) {
       private_key_method_ =
           factory_context.sslContextManager()
               .privateKeyMethodManager()
-              .createPrivateKeyMethodProvider(config.private_key_provider(), factory_context);
+              .createPrivateKeyMethodProvider(config.private_key_provider(), factory_context,
+                                              const_cast<std::string&>(private_key_));
+      if (private_key_method_ && !private_key_method_->checkInitialized()) {
+        private_key_method_ = nullptr;
+      }
     }
     if (certificate_chain_.empty()) {
       throw EnvoyException(
           fmt::format("Failed to load incomplete certificate from {}: certificate chain not set",
                       certificate_chain_path_));
     }
-    if (private_key_.empty() && private_key_method_ == nullptr) {
-      if (config.has_private_key_provider()) {
+
+    bool fallback = config.has_private_key_provider_list()
+                        ? config.private_key_provider_list().fallback()
+                        : true;
```

Comment:
> The default value of fallback should be false

I think the default value should be true, because if no private_key_provider_list config, software fallback should be enabled by default.

Reply:
If the default value is true, then that will change the API behavior. If we want to change the API behavior, then we need to add a runtime flag for it, but that is something we should discuss with the community this API change first.

My initial thought is that it isn't worth using a runtime flag for few cycles to deprecate the old behavior and bother the user for this API behavior change. That is why I set the default value as false.

Reply:
Second thoughts, the `private_key_provider_list` is new field, we can make the default fallback value as true. But we still need to ensure the fallback is false when only the old `private_key_provider` is specified

Reply:
I have splited the exception checkes for key provider(list) and private_key to avoid confusion with `fallback` variable.